Sources/NIOTestUtils/ManualTaskExecutor.swift

```diff
@@ -0,0 +1,164 @@
+///            It is highly recommended to use structured concurrency with these task executors.
+///
+/// - Parameters:
+///   - body: The closure that will accept the task executors.
+///
+/// - Throws: When `body` throws.
+///
+/// - Returns: The value returned by `body`.
+@available(macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0, *)
+@inlinable
+public func withManualTaskExecutor<T, Failure>(
+    body: (ManualTaskExecutor, ManualTaskExecutor) async throws(Failure) -> T
+) async throws(Failure) -> T {
+    let taskExecutor1 = ManualTaskExecutor()
+    defer { taskExecutor1.shutdown() }
+
+    let taskExecutor2 = ManualTaskExecutor()
+    defer { taskExecutor2.shutdown() }
+
+    return try await body(taskExecutor1, taskExecutor2)
+}
+
+/// Manual task executor.
+///
+/// A `TaskExecutor` that does not use any threadpool or similar mechanism to run the jobs.
+/// Jobs are manually run by calling the `runUntilQueueIsEmpty` method.
+///
+@available(macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0, *)
+public final class ManualTaskExecutor: TaskExecutor {
+    struct Storage {
+        var isShutdown = false
+        var jobs = Deque<UnownedJob>()
+    }
+
+    private let storage = Mutex<Storage>(.init())
+
+    @usableFromInline
+    init() {}
+
+    /// Run jobs until queue is empty.
+    ///
+    /// Synchronously runs all enqueued jobs, including any jobs that are enqueued while running.
+    /// When this function returns, it means that each task running on this executor is either:
+    /// - suspended
+    /// - moved (temporarily) to a different executor
+    /// - finished
+    ///
+    /// If not all tasks are finished, this function must be called again.
+    public func runUntilQueueIsEmpty() {
+        while let job = self.storage.withLock({ $0.jobs.popFirst() }) {
+            job.runSynchronously(on: self.asUnownedTaskExecutor())
+        }
+    }
+
+    /// Enqueue a job.
+    ///
+    /// Called by the concurrency runtime.
+    ///
+    /// - Parameter job: The job to enqueue.
+    public func enqueue(_ job: UnownedJob) {
+        self.storage.withLock { storage in
+            if storage.isShutdown {
+                fatalError("A job is enqueued after manual executor shutdown")
+            }
+            storage.jobs.append(job)
+        }
+    }
+
+    /// Shutdown.
+    ///
+    /// Since the manual task executor is not running anything in the background, this is purely to catch
+    /// any issues due to incorrect usage of the executor. The shutdown verifies that the queue is empty
+    /// and makes sure that no new jobs can be enqueued.
+    @usableFromInline
+    func shutdown() {
+        self.storage.withLock { storage in
+            if !storage.jobs.isEmpty {
+                fatalError("Shutdown of manual executor with jobs in queue")
+            }
+            storage.isShutdown = true
+        }
+    }
```

Comment:
I don't know if we really need this complicated shutdown handling. Since task executors can only be set on a scope we shouldn't have to do any manual clean up or reject jobs.

Reply:
It can catch some incorrect usage, like for example:
```swift
await withManualTaskExecutor { taskExecutor in
    Task(executorPreference: taskExecutor) {
        try await Task.sleep(for: .seconds(3))
    }
    taskExecutor.runUntilQueueIsEmpty()  // Runs until the sleep starts, but the task is not finished yet
}
```
It will definitely not catch all problems, but I was thinking there could be value in catching some.

Reply:
Since it is not a public thing anymore I'm fine with leaving it. @Lukasa @glbrntt you want to take a look before we merge it?